src/back/spv/writer.rs

```diff
@@ -885,10 +885,36 @@ impl Writer {
         output: &mut Vec<Instruction>,
     ) -> Instruction {
         match statement {
-            crate::Statement::Return { value: _ } => match function.return_type {
-                Some(_) => unimplemented!(),
-                None => Instruction::new(Op::Return),
-            },
+            crate::Statement::Return { value } => {
+                match function.return_type {
+                    Some(expected_ty) => {
+                        let value = value.expect(
+                            "Return with no value from function supposed to return a value",
+                        );
+
+                        // Parse the expression
+                        let value_expression = &function.expressions[value];
+                        let (value_id, value_ty) =
+                            self.parse_expression(ir_module, function, value_expression, output);
+
+                        // Check that the actual returned type matches the expected one
+                        let expected_ty = &ir_module.types[expected_ty].inner;
+                        assert_eq!(expected_ty, value_ty, "Type of returned expression does not match declared function return type");
+
+                        // Construct the return value instruction
+                        let mut instruction = Instruction::new(Op::ReturnValue);
+                        instruction.set_result(value_id);
+                        instruction
+                    }
+                    None => {
+                        assert!(
+                            value.is_none(),
+                            "Trying to return value from function returning nothing"
+                        );
```

Comment:
Same for this one, @kvark. This validation should be done at IR level too I think?

Reply:
Correct! The backend should *assume* that IR is valid. We don't want every backend doing its own validation.